compiler/rustc_middle/src/query/mod.rs

```diff
@@ -1871,8 +1871,40 @@ rustc_queries! {
 
     query subst_and_check_impossible_predicates(key: (DefId, SubstsRef<'tcx>)) -> bool {
         desc { |tcx|
-            "impossible substituted predicates:`{}`",
-            tcx.def_path_str(key.0)
+            "checking for impossible predicates for `{}` substituting `{:?}`",
+            tcx.def_path_str(key.0), key.1
+        }
+    }
+
+    // Check if it's even possible to satisfy the 'where' clauses
+    // for this item, without substitutions.
+    //
+    // This branch will never be taken for any normal function.
+    // However, it's possible to `#!feature(trivial_bounds)]` to write
+    // a function with impossible to satisfy clauses, e.g.:
+    // `fn foo() where String: Copy {}`
+    //
+    // We don't usually need to worry about this kind of case,
+    // since we would get a compilation error if the user tried
+    // to call it. However, since we can do certain mir optimizations
+    // even without any calls to the function, we need to make
+    // sure that it even makes sense to try to evaluate the body.
+    //
+    // We manually filter the predicates, skipping anything that's not
+    // "global". We are in a potentially generic context
+    // (e.g. we are evaluating a function without substituting generic
+    // parameters), so this filtering serves two purposes:
+    //
+    // 1. We skip evaluating any predicates that we would
+    // never be able prove are unsatisfiable (e.g. `<T as Foo>`
+    // 2. We avoid trying to normalize predicates involving generic
+    // parameters (e.g. `<T as Foo>::MyItem`). This can confuse
+    // the normalization code (leading to cycle errors), since
+    // it's usually never invoked in this way.
+    query check_impossible_predicates_for_item(key: DefId) -> bool {
```

Comment:
Could you rename the query to make the `bool` more explicit?
Something like `item_has_impossible_predicates`?